### PR TITLE
feat: check validity of trace- and span-id in context update from header

### PR DESCRIPTION
**Pull request by JoshuaMoelans**
Analysis of https://github.com/getsentry/sentry-native/issues/1012 showed a potential way for users to send invalid trace and span id values. This can happen when a customer uses `sentry_transaction_context_update_from_header(...)` which reads a string that looks like `traceid-spanid(-sampled)?`. The only formatting that currently gets checked is whether there is a first `-`.

Although these are never used in an unsafe context within our code, it still makes sense to reject the invalid formats on our end rather than [relying on Relay ](https://github.com/getsentry/relay/blob/master/relay-event-schema/src/protocol/contexts/trace.rs#L17-L21) to reject them.

The formats we expect are the following (from [our docs](https://develop.sentry.dev/sdk/telemetry/traces/#header-sentry-trace)):
- `trace_id` = 32 hex characters
- `span_id` = 16 hex characters

ToDo
- [x] Make sure we agree on this approach
- [x] Add a warning mechanism when an invalid input is received
- [x] Update the relevant [docs page](https://docs.sentry.io/platforms/native/tracing/trace-propagation/custom-instrumentation/)

### CODE REVIEW

**Comment by JoshuaMoelans** [comment]
ToDo: update the docs (think it is [this page](https://docs.sentry.io/platforms/native/tracing/trace-propagation/custom-instrumentation/)) with a valid header.

File: src/sentry_tracing.c [hunk 1]
```diff
@@ -151,6 +154,26 @@ sentry_transaction_context_remove_sampled(sentry_transaction_context_t *tx_cxt)
...
+bool
+is_valid_nonzero_hexstring(const char *s, size_t len)
+{
+    bool has_nonzero = false;
+    for (size_t i = 0; i < len; i++) {
+        if (!isxdigit(s[i])) {
+            return false;
+        }
+        if (s[i] != '0') {
+            has_nonzero = true;
```
Comment (supervacuus) [inline src/sentry_tracing.c:171]: If the condition is to have _at least one_ non-zero character, you could early return here as well.

**Review by supervacuus — CHANGES_REQUESTED** [review]

File: src/sentry_tracing.c (hunk 1 above)
Comment (JoshuaMoelans) [reply src/sentry_tracing.c:171]: I think we want at least one non-zero character **and** for all characters to be hexadecimal, so seeing one non-zero character isn't enough (e.g. we could have `00000000ABCZZZZZ` where Z makes it an invalid span-id).

File: src/sentry_tracing.c [hunk 2]
```diff
@@ -151,6 +154,26 @@ sentry_transaction_context_remove_sampled(sentry_transaction_context_t *tx_cxt)
     }
 }
 
+///
+/// @param s given string
+/// @param len the length of the string
+/// @return whether the string is a valid hex string of at least the given
+///         length which contains at least one non-zero character
```
[the remark this answers is not in the document]
Comment (JoshuaMoelans) [reply src/sentry_tracing.c]: Makes sense! Updated in commit https://github.com/getsentry/sentry-native/pull/1046/commits/dabcfca6f0023b2e50b2ab863f96be3b7cefa048

File: src/sentry_tracing.c [hunk 3]
```diff
@@ -8,6 +8,9 @@
 #include "sentry_value.h"
 #include <string.h>
 
+#define TRACE_ID_LEN 32
+#define SPAN_ID_LEN 16
```
[the remark this answers is not in the document]
Comment (JoshuaMoelans) [reply src/sentry_tracing.c]: Thanks, this is indeed cleaner & more readable. Updated this in commit https://github.com/getsentry/sentry-native/pull/1046/commits/dabcfca6f0023b2e50b2ab863f96be3b7cefa048 with a function that can check either the trace or span id by having the expected length as a parameter.

File: src/sentry_tracing.c [hunk 4]
```diff
@@ -151,6 +154,26 @@ sentry_transaction_context_remove_sampled(sentry_transaction_context_t *tx_cxt)
     }
 }
 
+///
+/// @param s given string
+/// @param len the length of the string
+/// @return whether the string is a valid hex string of at least the given
+///         length which contains at least one non-zero character
+bool
+is_valid_nonzero_hexstring(const char *s, size_t len)
```
[the remark this answers is not in the document]
Comment (JoshuaMoelans) [reply src/sentry_tracing.c]: Updated in commit https://github.com/getsentry/sentry-native/pull/1046/commits/dabcfca6f0023b2e50b2ab863f96be3b7cefa048

**Review by supervacuus — APPROVED** [review]

**Comment by supervacuus** [comment]
I went ahead with the merge and created a docs issue as a follow-up to your todo item: https://github.com/getsentry/sentry-docs/issues/11504

**Pull request merged**